### PR TITLE
Upload arweave function

### PR DESCRIPTION
**Pull request by lyfer233**
Added:
1. Using the third package `arweave-rs` connect with the arweave.
2. Added a new field `arweave_id` to the kv_chains table and update the `src/schema.rs` file.
3. Updated two unit tests because of adding a field.
4. Add a unit test to ensure the connection with arweave and upload function.
5. Error process.

### CODE REVIEW

File: migrations/2023_08_01-add_arewave_id_to_kv_chains/up.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+-- Your SQL goes here
+
+ALTER TABLE kv_chains
+ADD arweave_id VARCHAR DEFAULT '';
```
Comment (nykma) [inline migrations/2023_08_01-add_arewave_id_to_kv_chains/up.sql]: Will `DEFAULT ''` be conflict with `Option<String>` ? I'm not sure about this.

File: src/controller/upload.rs [hunk 2]
```diff
@@ -47,8 +48,28 @@ pub async fn controller(request: Request) -> Result<Response, Error> {
     // Validate signature
     new_kv.validate()?;
 
+    // Try take the kvchain data upload to the arweave.
+    let arweave_document = KVChainArweaveDocument{
+        avatar: avatar.unwrap(),
```
Comment (nykma) [inline src/controller/upload.rs]: Will `unwrap()` here be paniced?

File: src/controller/upload.rs [hunk 3]
```diff
@@ -47,8 +48,28 @@ pub async fn controller(request: Request) -> Result<Response, Error> {
...
+        platform: new_kv.platform.clone(),
+        identity: new_kv.identity.clone(),
+        patch: new_kv.patch.clone(),
+        signature: new_kv.signature.clone(),
+        created_at: new_kv.created_at,
+        signature_payload: new_kv.signature_payload.clone(),
+        previous_uuid: None,
+    };
+    
+    new_kv.arweave_id = match arweave_document.upload_to_arweave().await {
```
Comment (nykma) [inline src/controller/upload.rs]:
```suggestion
    new_kv.arweave_id = arweave_document.upload_to_arweave().ok();
```

File: src/error/mod.rs [hunk 4]
```diff
@@ -52,6 +58,8 @@ impl Error {
             Error::SignatureValidationError(_) => StatusCode::BAD_REQUEST,
             Error::Base64Error(_) => StatusCode::BAD_REQUEST,
             Error::UuidParseError(_) => StatusCode::BAD_REQUEST,
+            Error::UrlParseError(_) => StatusCode::BAD_REQUEST,
+            Error::ArweaveError(_) => StatusCode::BAD_REQUEST,
```
Comment (nykma) [inline src/error/mod.rs]: Basiclly Arweave upload error is not user's fault, so it should be `5xx`, like `INTERNAL_SERVER_ERROR`

File: src/model/arweave/mod.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+use ::uuid::Uuid;
+use chrono::NaiveDateTime;
+use arweave_rs::{Arweave, crypto::base64::Base64};
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+use crate::error::Error;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct KVChainArweaveDocument {
```
Comment (nykma) [inline src/model/arweave/mod.rs:16]: Why creating a new `Model` here?

**Review by nykma — CHANGES_REQUESTED** [review]

**Comment by lyfer233** [comment]
Current problem:
- `arweave_id` not normally add into response when user post to `/v1/kv`, so I will think how to design and refactor some codes to address this issue.

File: src/controller/upload.rs [hunk 6]
```diff
@@ -47,12 +48,30 @@ pub async fn controller(request: Request) -> Result<Response, Error> {
...
+        platform: new_kv.platform.clone(),
+        identity: new_kv.identity.clone(),
+        patch: new_kv.patch.clone(),
+        signature: new_kv.signature.clone(),
+        created_at: new_kv.created_at,
+        signature_payload: new_kv.signature_payload.clone(),
+        previous_uuid: None,
+    };
+    
+    new_kv.arweave_id = arweave_document.upload_to_arweave().await.ok();
```
Comment (nykma) [inline src/controller/upload.rs]: Please make this upload procedure a "background job". 
Uploading should not "block" this upload request.

**Comment by lyfer233** [comment]
Next plans:
1. fix this problem https://github.com/NextDotID/kv_server/pull/12#issuecomment-1705489400
2. Use thread to run upload job
3. Add more unit test to ensure error process and corner condition.

File: src/controller/upload.rs [hunk 7]
```diff
@@ -47,13 +48,38 @@ pub async fn controller(request: Request) -> Result<Response, Error> {
...
     let kv_link = new_kv.finalize(&mut conn)?;
+
     // Apply patch
-    kv_link.perform_patch(&mut conn)?;
+    kv_link.perform_patch(&mut conn, new_kv.arweave_id)?;
 
     // All done. Build response.
     let response = query_response(&mut conn, &persona.public_key)?;
+
+    new_kv.arweave_id = arweave_upload_future.await.unwrap();
```
Comment (nykma) [inline src/controller/upload.rs]: It still blocks `json_response` below.

What I mean is, spawning a tokio background job and leave it alone.

**Review by nykma — CHANGES_REQUESTED** [review]
Good enough. Only a small thing to tweak.

**Review by nykma — COMMENTED** [review]
Good for now. I'll fix some trivial stuff.

**Review by nykma — APPROVED** [review]
Good for now. I'll fix some trivial stuff.
